harness/determined/pytorch/_trainer.py

```diff
@@ -0,0 +1,233 @@
+            local_rank=self._core.distributed.get_local_rank(),
+            begin_on_batch=begin_on_batch,
+            end_after_batch=end_after_batch,
+            sync_timings=sync_timings,
+        )
+
+    def fit(
+        self,
+        checkpoint_period: Optional[pytorch.TrainUnit] = None,
+        validation_period: Optional[pytorch.TrainUnit] = None,
+        max_length: Optional[pytorch.TrainUnit] = None,
+        reporting_period: Optional[pytorch.TrainUnit] = None,
+        aggregation_frequency: Optional[int] = None,
+        checkpoint_policy: Optional[str] = None,
+        test_mode: Optional[bool] = None,
+    ) -> None:
+
+        # Set context and training variables
+        if aggregation_frequency:
+            self._context._aggregation_frequency = aggregation_frequency
+
+        # Set defaults
+        checkpoint_policy = checkpoint_policy or "best"
+        checkpoint_period = checkpoint_period or pytorch.Batch(sys.maxsize)
+        validation_period = validation_period or pytorch.Batch(sys.maxsize)
+        test_mode = test_mode or False
+
+        if self._local_training:
+            if checkpoint_policy == "best":
+                logging.warning(
+                    "checkpoint_policy='best' is not supported in local training mode. "
+                    "Falling back to 'all'"
+                )
+                checkpoint_policy = "all"
+            assert max_length, "max_length must be defined in local training mode"
+            assert isinstance(
+                max_length.value, int
+            ), "max_length must be configured in TrainUnit(int)"
+
+            if self._det_profiler:
+                logging.warning("Determined profiler will be ignored in local training mode")
+
+            latest_checkpoint = None
+            smaller_is_better = True
+            searcher_metric_name = None
+            steps_completed = 0
+            reporting_period = reporting_period or pytorch.Batch(sys.maxsize)
+            step_zero_validation = False
+            global_batch_size = None
+        else:
+
+            assert not test_mode, "test_mode is only supported in local training mode"
+            assert self._info, "Unable to detect cluster info"
+            if max_length is not None:
+                logging.warning(
+                    "max_length is ignored when training on-cluster. Please configure the searcher instead"
+                )
+
+            latest_checkpoint = self._info.latest_checkpoint
+            smaller_is_better = bool(self._info.trial._config["searcher"]["smaller_is_better"])
+
+            searcher_metric_name = self._info.trial._config["searcher"]["metric"]
+            steps_completed = int(self._info.trial._steps_completed)
+            reporting_period = reporting_period or pytorch.Batch(
+                int(self._info.trial._config["scheduling_unit"])
+            )
+            step_zero_validation = bool(self._info.trial._config["perform_initial_validation"])
```

Comment:
There's a lot of backwards-compatibility steps happening in here, can we get rid of all of it?

I don't see why `step_zero_validation` would be read from the experiment config for a user using the trainer api directly.  It should just be an argument to fit.

Honestly I think anything in this function that is reading from `self._info` is suspect...